futures-async-macro/README.md

```diff
@@ -0,0 +1,51 @@
+
+**Note that these are experimental APIs.**
+
+## Usage
+
+Add this to your `Cargo.toml`:
+
+```toml
+[dependencies]
+futures-preview = { version = "0.3.0-alpha.14", features = ["async-stream", "nightly"] }
+```
+
+### \#\[for_await\]
+
+Processes streams using a for loop.
+
+This is a reimplement of [futures-await]'s `#[async]` for loops for futures 0.3 and is an experimental implementation of [the idea listed as the next step of async/await](https://github.com/rust-lang/rfcs/blob/master/text/2394-async_await.md#for-await-and-processing-streams).
+
+```rust
+use futures::for_await;
+use futures::prelude::*;
+
+#[for_await]
+for value in stream::iter(1..=5) {
+    println!("{}", value);
+}
+```
+
+`value` has the `Item` type of the stream passed in. Note that async for loops can only be used inside of `async` functions or `#[async_stream]` functions.
+
+### \#\[async_stream\]
+
+This is a reimplement of [futures-await]'s `#[async_stream]` for futures 0.3 and is an experimental implementation of [the idea listed as the next step of async/await](https://github.com/rust-lang/rfcs/blob/master/text/2394-async_await.md#generators-and-streams).
+
+```rust
+use futures::prelude::*;
+use futures::async_stream;
+
+// Returns a stream of i32
+#[async_stream]
+fn foo(stream: impl Stream<Item = String>) -> i32 {
```

Comment:
One of the concerns with current implementations is that there is an Item type at the "return" position. In the `futures-await` implementation this was specified in the form of an attribute argument `#[async_stream(Item = i32)]`.

Reply:
@taiki-e that approach seems reasonable to me; are there any reasons why we can't / shouldn't follow `future-await`'s approach?

Reply:
@yoshuawuyts There is no particular reason. Also, it is easy to change the implementation to that.

(Because the stream has no final return value, the return position is always empty. The current implementation is just trying to reuse this blank.)

Reply:
In [futures-await](https://github.com/alexcrichton/futures-await) we can use `?` operator inside stream. If I understand correctly, with new syntax it's impossible. Is it true?

```rust
/// Fetch all provided urls one at a time
#[async_stream(item = String)]
fn fetch_all(client: hyper::Client, urls: Vec<&'static str>) -> io::Result<()> {
    for url in urls {
        let s = await!(fetch(client, url))?;
        stream_yield!(s);
    }
    Ok(())
}
```

Reply:
@kgv yes, because of streams no longer being required to have an error and item type that no longer works.

You can probably do something mostly equivalent with a `try` block though:

```rust
/// Fetch all provided urls one at a time
#[async_stream(item = io::Result<String>)]
fn fetch_all(client: hyper::Client, urls: Vec<&'static str>) {
    let res = try {
        for url in urls {
            yield Ok(fetch(client, url).await?);
        }
    };
    if let Err(err) = res {
         yield Err(err);
    }
}
```

and it may be possible to add more sugar for this in the future (an `async_try_stream` maybe?).

Reply:
@Nemo157
> and it may be possible to add more sugar for this in the future (an `async_try_stream` maybe?).

I will try it after this PR is merged. And I think it will be something like the following:

```rust
#[async_try_stream(Ok = String, Error = io::Error)]
fn fetch_all(client: hyper::Client, urls: Vec<&'static str>) {
    for url in urls {
         yield fetch(client, url).await?;
    }
}

// desugaring

fn fetch_all(client: hyper::Client, urls: Vec<&'static str>) -> impl Stream<Item = Result<String, io::Error>> {
    from_generator(static move || { // impl Generator<Yield = String, Return = Result<(), io::Error>>
        for url in urls {
             yield Poll::Ready(fetch(client, url).await?);
        }
        return Ok(()); // It is the same behavior as try_blocks that this is added automatically.
        loop { yield Poll::Pending }
    })
}

```

Reply:
(Also, if we add `async_try_stream`, it may be preferable to write `Item` at the position of the attribute argument, not at the position of return, so I will revert it)